Perf benchmark test runner fixes

carolynhu opened the pull request.

istio-testing commented:
@carolynhu: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
tools-build | 0f2f5d1a01cf8d1050ddf40fb41ce90eb62684d4 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/300/tools-build/91) | `/test tools-build`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@carolynhu: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
tools-build | 0f2f5d1a01cf8d1050ddf40fb41ce90eb62684d4 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/300/tools-build/91) | `/test tools-build`
tools-test | 0f2f5d1a01cf8d1050ddf40fb41ce90eb62684d4 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/300/tools-test/92) | `/test tools-test`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented on perf/benchmark/README.md:
```diff
@@ -86,77 +86,81 @@ Required fields:
+NAME           TYPE           CLUSTER-IP   EXTERNAL-IP     PORT(S)                                                       AGE
+fortioclient   LoadBalancer   10.0.28.56   35.225.105.18   8080:31759/TCP,8079:30495/TCP,8078:31107/TCP,8077:31034/TCP   16h
+
+$ export FORTIO_CLIENT_URL=http://35.225.105.18:8080
```
probably don't want to commit your public ip address here.

We could replace with 

`export FORTIO_CLIENT_URL=$(kubectl get service fortioclient -o jsonpath='{.status.loadBalancer.ingress[0].ip}'):8080`

maybe

istio-testing commented:
@carolynhu: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
tools-test | 0f2f5d1a01cf8d1050ddf40fb41ce90eb62684d4 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/300/tools-test/92) | `/test tools-test`
tools-build | 7f25025c295738a89b8ca2a1febcd1b53755000d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/300/tools-build/92) | `/test tools-build`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@carolynhu: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
tools-build | 7f25025c295738a89b8ca2a1febcd1b53755000d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/300/tools-build/92) | `/test tools-build`
tools-test | 7f25025c295738a89b8ca2a1febcd1b53755000d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/300/tools-test/93) | `/test tools-test`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@carolynhu: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
tools-test | 7f25025c295738a89b8ca2a1febcd1b53755000d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/300/tools-test/93) | `/test tools-test`
tools-build | 7ff206d210e93424467dc450c0f741a2f19af857 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/300/tools-build/93) | `/test tools-build`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@carolynhu: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
tools-build | 7ff206d210e93424467dc450c0f741a2f19af857 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/300/tools-build/93) | `/test tools-build`
tools-test | 7ff206d210e93424467dc450c0f741a2f19af857 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/300/tools-test/94) | `/test tools-test`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mandarjog commented:
@carolynhu what is exactly broken in the runner? Can you add the error messages here or to the attached issue?

carolynhu commented:
Here are the errors:
(1) When I run `python runner/runner.py <conn> <qps> <duration> --OPTIONAL-FLAGS` with  `--perf` . The kubecp() func throws error:
```
kubectl --namespace {namespace} cp {from_file} {to_file} -c istio-proxy
error: one of src or dest must be a remote file specification
Traceback (most recent call last):
  File "runner/runner.py", line 359, in <module>
    sys.exit(main(sys.argv[1:]))
  File "runner/runner.py", line 354, in main
    return run(args)
  File "runner/runner.py", line 284, in run
    fortio.run(conn=conn, qps=qps, duration=args.duration, size=args.size)
  File "runner/runner.py", line 199, in run
    duration=40)
  File "runner/runner.py", line 219, in perf
    kubecp(mesh, PERFSH, pod + ":" + perfpath)
  File "runner/runner.py", line 243, in kubecp
    return run_command_sync(cmd)
  File "runner/runner.py", line 38, in run_command_sync
    op = subprocess.check_output(command, shell=True)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/subprocess.py", line 395, in check_output
    **kwargs).stdout
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/subprocess.py", line 487, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command 'kubectl --namespace {namespace} cp {from_file} {to_file} -c istio-proxy' returned non-zero exit status 1.
```
(2) Without `--perf` command in the last step, I can successfully run all those tests. Then, when I run this `python ./runner/fortio.py $FORTIO_CLIENT_URL --prometheus=$PROMETHEUS_URL --csv .....` to gather the results. I got timeout error:
```Fetching prometheus metrics for 2019-08-23 16:05:08 192a55b6_qps_1000_c_4_1024_both
... 192a55b6_qps_1000_c_4_1024_both duration=60s is less than minimum 92s
Fetching prometheus metrics for 2019-08-23 16:04:07 192a55b6_qps_1000_c_4_1024_serveronly
... 192a55b6_qps_1000_c_4_1024_serveronly duration=60s is less than minimum 92s
Fetching prometheus metrics for 2019-08-23 16:01:35 443fbb69_qps_1000_c_16_1024_both
Traceback (most recent call last):
  File "/Users/carolynprh/.local/share/virtualenvs/runner-xV1YZQFd/lib/python3.7/site-packages/urllib3/connection.py", line 160, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw)
  File "/Users/carolynprh/.local/share/virtualenvs/runner-xV1YZQFd/lib/python3.7/site-packages/urllib3/util/connection.py", line 80, in create_connection
    raise err
  File "/Users/carolynprh/.local/share/virtualenvs/runner-xV1YZQFd/lib/python3.7/site-packages/urllib3/util/connection.py", line 70, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 61] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/carolynprh/.local/share/virtualenvs/runner-xV1YZQFd/lib/python3.7/site-packages/urllib3/connectionpool.py", line 603, in urlopen
    chunked=chunked)
  File "/Users/carolynprh/.local/share/virtualenvs/runner-xV1YZQFd/lib/python3.7/site-packages/urllib3/connectionpool.py", line 355, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1244, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1290, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1239, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 966, in send
    self.connect()
  File "/Users/carolynprh/.local/share/virtualenvs/runner-xV1YZQFd/lib/python3.7/site-packages/urllib3/connection.py", line 183, in connect
    conn = self._new_conn()
  File "/Users/carolynprh/.local/share/virtualenvs/runner-xV1YZQFd/lib/python3.7/site-packages/urllib3/connection.py", line 169, in _new_conn
    self, "Failed to establish a new connection: %s" % e)
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x109dc1b90>: Failed to establish a new connection: [Errno 61] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/carolynprh/.local/share/virtualenvs/runner-xV1YZQFd/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/Users/carolynprh/.local/share/virtualenvs/runner-xV1YZQFd/lib/python3.7/site-packages/urllib3/connectionpool.py", line 641, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/Users/carolynprh/.local/share/virtualenvs/runner-xV1YZQFd/lib/python3.7/site-packages/urllib3/util/retry.py", line 399, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=9090): Max retries exceeded with url: /api/v1/query_range?query=irate%28container_cpu_usage_seconds_total%7Bjob%3D%22kubernetes-cadvisor%22%2Ccontainer_name%3D~%22mixer%7Cpolicy%7Cdiscovery%7Cistio-proxy%7Ccaptured%7Cuncaptured%22%7D%5B1m%5D%29&start=1566576157&end=1566576182&step=15 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x109dc1b90>: Failed to establish a new connection: [Errno 61] Connection refused'))
```
(3) Even if I have the cvs at hand, like I converted the one of the json file at this repo https://github.com/istio/tools/tree/master/perf/benchmark/results-istio-1.2.0-fortio, and run graph.py to generate the graph. That script seems not able to handle float data type properly.
```
(runner) bash-3.2$ python graph.py 2019-07-10-165728_193ff770_qps_1000_c_1_1024_serveronly.csv p50
Generating chart, x_label=connections, y_label=p50, csv=2019-07-10-165728_193ff770_qps_1000_c_1_1024_serveronly.csv ...
Traceback (most recent call last):
  File "graph.py", line 186, in <module>
    sys.exit(main(sys.argv[1:]))
  File "graph.py", line 162, in main
    return generate_chart(args.mesh, args.csv, args.xaxis, args.metric)
  File "graph.py", line 39, in generate_chart
    df["Labels"] = [x.split('_', 6)[-1] for x in df['Labels']]
  File "graph.py", line 39, in <listcomp>
    df["Labels"] = [x.split('_', 6)[-1] for x in df['Labels']]
AttributeError: 'float' object has no attribute 'split'
```

istio-testing commented:
@carolynhu: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>
